{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standard variables loaded, you are good to go!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "from os.path import join, dirname, basename, exists, isdir\n",
    "\n",
    "### Load environmental variables from the project root directory ###\n",
    "# find .env automagically by walking up directories until it's found\n",
    "dotenv_path = find_dotenv()\n",
    "\n",
    "# load up the entries as environment variables\n",
    "load_dotenv(dotenv_path)\n",
    "\n",
    "# now you can get the variables using their names\n",
    "\n",
    "# Check whether a network drive has been specified\n",
    "DATABASE = os.environ.get(\"NETWORK_URL\")\n",
    "if DATABASE == 'None':\n",
    "    pass\n",
    "else:\n",
    "    pass\n",
    "    #mount network drive here\n",
    "\n",
    "# set up directory pathsa\n",
    "CURRENT_DIR = os.getcwd()\n",
    "PROJ = dirname(dotenv_path) # project root directory\n",
    "\n",
    "DATA = join(PROJ, 'data') #data directory\n",
    "RAW_EXTERNAL = join(DATA, 'raw_external') # external data raw directory\n",
    "RAW_INTERNAL = join(DATA, 'raw_internal') # internal data raw directory\n",
    "INTERMEDIATE = join(DATA, 'intermediate') # intermediate data directory\n",
    "FINAL = join(DATA, 'final') # final data directory\n",
    "\n",
    "RESULTS = join(PROJ, 'results') # output directory\n",
    "FIGURES = join(RESULTS, 'figures') # figure output directory\n",
    "PICTURES = join(RESULTS, 'pictures') # picture output directory\n",
    "\n",
    "\n",
    "# make folders specific for certain data\n",
    "folder_name = ''\n",
    "if folder_name != '':\n",
    "    #make folders if they don't exist\n",
    "    if not exists(join(RAW_EXTERNAL, folder_name)):\n",
    "        os.makedirs(join(RAW_EXTERNAL, folder_name))\n",
    "\n",
    "    if not exists(join(INTERMEDIATE, folder_name)):\n",
    "        os.makedirs(join(INTERMEDIATE, folder_name))\n",
    "\n",
    "    if not exists(join(FINAL, folder_name)):\n",
    "        os.makedirs(join(FINAL, folder_name))\n",
    "\n",
    "\n",
    "print('Standard variables loaded, you are good to go!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "# import data \n",
    "data = pd.read_csv(f\"{INTERMEDIATE}/proteomics/acetate_usages.csv\", index_col=0)\n",
    "\n",
    "data.head()\n",
    "\n",
    "# get cell volumes\n",
    "cell_volumes = pd.read_csv(f\"{RAW_INTERNAL}/proteomics/growth_conditions.csv\", index_col=0)\n",
    "cell_volumes = cell_volumes[\"Single cell volume [fl]1\"]\n",
    "# remove the first two rows of LB\n",
    "cell_volumes = cell_volumes.loc[~cell_volumes.index.duplicated(keep='first')]\n",
    "# rename the number 3 in there\n",
    "cell_volumes = cell_volumes.rename({'Osmotic-stress glucose3':'Osmotic-stress glucose_uncertainty'}, axis='index')\n",
    "rename_dict = {i:re.sub(r'\\W+', '', i).lower() for i in cell_volumes.index}\n",
    "cell_volumes = cell_volumes.rename(rename_dict, axis='index')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ac_1</th>\n",
       "      <th>ac_2</th>\n",
       "      <th>ac_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O32583</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00350</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00363</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00370</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00393</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ac_1  ac_2  ac_3\n",
       "O32583   0.0   0.0   0.0\n",
       "P00350   0.0   0.0   0.0\n",
       "P00363   0.0   0.0   0.0\n",
       "P00370   0.0   0.0   0.0\n",
       "P00393   0.0   0.0   0.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Finally, convert to mmol/gDW:\n",
    "water_content = 0.3\n",
    "cell_density = 1.105e-12\n",
    "\n",
    "\n",
    "\n",
    "# Iterate through the dataset and multiply by the corresponding cell volume, to get mmol/fL:\n",
    "for (col_name, d) in data.iteritems():\n",
    "    chemo_name = \"acetate\"\n",
    "    try:\n",
    "        data[col_name] = data[col_name] * cell_volumes.loc[chemo_name]\n",
    "    except:\n",
    "        print(chemo_name)\n",
    "        \n",
    "        \n",
    "data = data * cell_density * water_content\n",
    "        \n",
    "\n",
    "# convert into counts\n",
    "data = data * 6.022e+23 / 1000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_for_UP2genes = pd.read_csv(f\"{RAW_EXTERNAL}/raw_proteomics_all.csv\")\n",
    "data_for_UP2genes = data_for_UP2genes.drop([0])\n",
    "UP2gene = dict(zip(data_for_UP2genes[\"UP\"], data_for_UP2genes[\"Frist_Gene\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate index uniprot names to gene names and remove enzymes not included in the data \n",
    "data = data.loc[list(set(data.index).intersection(UP2gene.keys()))]\n",
    "data.index = [UP2gene[i] for i in list(set(data.index).intersection(UP2gene.keys()))]\n",
    "#data[\"pyr_1_original\"] = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ac_1</th>\n",
       "      <th>ac_2</th>\n",
       "      <th>ac_3</th>\n",
       "      <th>ac_original</th>\n",
       "      <th>ac_original_1</th>\n",
       "      <th>ac_original_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chbB</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>778.52347098222</td>\n",
       "      <td>722.842713843346</td>\n",
       "      <td>643.763656524197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yghU</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>995.133432621134</td>\n",
       "      <td>1040.64573696801</td>\n",
       "      <td>927.386633178478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glpA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.3096198203812</td>\n",
       "      <td>12.5791009044582</td>\n",
       "      <td>16.8251266013508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rffG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.116024889242</td>\n",
       "      <td>13.1699483088472</td>\n",
       "      <td>14.6052208502821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acnB</th>\n",
       "      <td>38842.730483</td>\n",
       "      <td>38842.730483</td>\n",
       "      <td>38842.730483</td>\n",
       "      <td>22426.9657144047</td>\n",
       "      <td>25951.6240916731</td>\n",
       "      <td>22843.9674488163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lpxD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36.3124976737935</td>\n",
       "      <td>48.5688880622485</td>\n",
       "      <td>48.0793517713787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ggt</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>182.562525556641</td>\n",
       "      <td>135.380378182929</td>\n",
       "      <td>152.896653173984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aceE</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2292.42314647829</td>\n",
       "      <td>2736.26880518675</td>\n",
       "      <td>2445.225832105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yqaB</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>52.5929585078648</td>\n",
       "      <td>64.8631209087853</td>\n",
       "      <td>31.5812142132445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>39.4708308323084</td>\n",
       "      <td>30.6687991970952</td>\n",
       "      <td>35.1031399768644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>855 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ac_1          ac_2          ac_3       ac_original  \\\n",
       "chbB      0.000000      0.000000      0.000000   778.52347098222   \n",
       "yghU      0.000000      0.000000      0.000000  995.133432621134   \n",
       "glpA      0.000000      0.000000      0.000000  15.3096198203812   \n",
       "rffG      0.000000      0.000000      0.000000   17.116024889242   \n",
       "acnB  38842.730483  38842.730483  38842.730483  22426.9657144047   \n",
       "...            ...           ...           ...               ...   \n",
       "lpxD      0.000000      0.000000      0.000000  36.3124976737935   \n",
       "ggt       0.000000      0.000000      0.000000  182.562525556641   \n",
       "aceE      0.000000      0.000000      0.000000  2292.42314647829   \n",
       "yqaB      0.000000      0.000000      0.000000  52.5929585078648   \n",
       "patA      0.000000      0.000000      0.000000  39.4708308323084   \n",
       "\n",
       "         ac_original_1     ac_original_2  \n",
       "chbB  722.842713843346  643.763656524197  \n",
       "yghU  1040.64573696801  927.386633178478  \n",
       "glpA  12.5791009044582  16.8251266013508  \n",
       "rffG  13.1699483088472  14.6052208502821  \n",
       "acnB  25951.6240916731  22843.9674488163  \n",
       "...                ...               ...  \n",
       "lpxD  48.5688880622485  48.0793517713787  \n",
       "ggt   135.380378182929  152.896653173984  \n",
       "aceE  2736.26880518675    2445.225832105  \n",
       "yqaB  64.8631209087853  31.5812142132445  \n",
       "patA  30.6687991970952  35.1031399768644  \n",
       "\n",
       "[855 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_for_UP2genes.index = data_for_UP2genes['Frist_Gene']\n",
    "\n",
    "data_for_UP2genes = data_for_UP2genes.loc[~data_for_UP2genes.index.duplicated(keep='first')]\n",
    "data = data.loc[~data.index.duplicated(keep='first')]\n",
    "data['ac_original'] = data_for_UP2genes.loc[data.index, ['acetate']]\n",
    "data['ac_original_1'] = data_for_UP2genes.loc[data.index, ['acetate.1']]\n",
    "data['ac_original_2'] = data_for_UP2genes.loc[data.index, ['acetate.2']]\n",
    "\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv(f\"{INTERMEDIATE}/proteomics/acetate_usages_counts.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
