{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standard variables loaded, you are good to go!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "from os.path import join, dirname, basename, exists, isdir\n",
    "\n",
    "### Load environmental variables from the project root directory ###\n",
    "# find .env automagically by walking up directories until it's found\n",
    "dotenv_path = find_dotenv()\n",
    "\n",
    "# load up the entries as environment variables\n",
    "load_dotenv(dotenv_path)\n",
    "\n",
    "# now you can get the variables using their names\n",
    "\n",
    "# Check whether a network drive has been specified\n",
    "DATABASE = os.environ.get(\"NETWORK_URL\")\n",
    "if DATABASE == 'None':\n",
    "    pass\n",
    "else:\n",
    "    pass\n",
    "    #mount network drive here\n",
    "\n",
    "# set up directory pathsa\n",
    "CURRENT_DIR = os.getcwd()\n",
    "PROJ = dirname(dotenv_path) # project root directory\n",
    "\n",
    "DATA = join(PROJ, 'data') #data directory\n",
    "RAW_EXTERNAL = join(DATA, 'raw_external') # external data raw directory\n",
    "RAW_INTERNAL = join(DATA, 'raw_internal') # internal data raw directory\n",
    "INTERMEDIATE = join(DATA, 'intermediate') # intermediate data directory\n",
    "FINAL = join(DATA, 'final') # final data directory\n",
    "\n",
    "RESULTS = join(PROJ, 'results') # output directory\n",
    "FIGURES = join(RESULTS, 'figures') # figure output directory\n",
    "PICTURES = join(RESULTS, 'pictures') # picture output directory\n",
    "\n",
    "\n",
    "# make folders specific for certain data\n",
    "folder_name = ''\n",
    "if folder_name != '':\n",
    "    #make folders if they don't exist\n",
    "    if not exists(join(RAW_EXTERNAL, folder_name)):\n",
    "        os.makedirs(join(RAW_EXTERNAL, folder_name))\n",
    "\n",
    "    if not exists(join(INTERMEDIATE, folder_name)):\n",
    "        os.makedirs(join(INTERMEDIATE, folder_name))\n",
    "\n",
    "    if not exists(join(FINAL, folder_name)):\n",
    "        os.makedirs(join(FINAL, folder_name))\n",
    "\n",
    "\n",
    "print('Standard variables loaded, you are good to go!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "# import data \n",
    "data = pd.read_csv(f\"{INTERMEDIATE}/proteomics/acetate_usages.csv\", index_col=0)\n",
    "\n",
    "\n",
    "\n",
    "# get cell volumes\n",
    "cell_volumes = pd.read_csv(f\"{RAW_INTERNAL}/proteomics/growth_conditions.csv\", index_col=0)\n",
    "cell_volumes = cell_volumes[\"Single cell volume [fl]1\"]\n",
    "# remove the first two rows of LB\n",
    "cell_volumes = cell_volumes.loc[~cell_volumes.index.duplicated(keep='first')]\n",
    "# rename the number 3 in there\n",
    "cell_volumes = cell_volumes.rename({'Osmotic-stress glucose3':'Osmotic-stress glucose_uncertainty'}, axis='index')\n",
    "rename_dict = {i:re.sub(r'\\W+', '', i).lower() for i in cell_volumes.index}\n",
    "cell_volumes = cell_volumes.rename(rename_dict, axis='index')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Growth condition\n",
       "lb                                  4.29\n",
       "glycerolaa                          3.83\n",
       "acetate                             2.30\n",
       "fumarate                            2.54\n",
       "galactose                           2.21\n",
       "glucose                             2.84\n",
       "glucosamine                         2.62\n",
       "glycerol                            2.64\n",
       "pyruvate                            2.50\n",
       "succinate                           2.58\n",
       "fructose                            2.96\n",
       "mannose                             2.64\n",
       "xylose                              2.79\n",
       "osmoticstressglucose_uncertainty    2.79\n",
       "42cglucose                          2.98\n",
       "ph6glucose                          2.93\n",
       "stationaryphase1day                 1.60\n",
       "stationaryphase3days                1.60\n",
       "chemostatµ012                       1.90\n",
       "chemostatµ020                       2.08\n",
       "chemostatµ035                       2.40\n",
       "chemostatµ05                        2.69\n",
       "Name: Single cell volume [fl]1, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_volumes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Finally, convert to mmol/gDW:\n",
    "water_content = 0.3\n",
    "cell_density = 1.105e-12\n",
    "\n",
    "\n",
    "\n",
    "# Iterate through the dataset and multiply by the corresponding cell volume, to get mmol/fL:\n",
    "for (col_name, d) in data.iteritems():\n",
    "    chemo_name = \"acetate\"\n",
    "    try:\n",
    "        data[col_name] = data[col_name] * cell_volumes.loc[chemo_name]\n",
    "    except:\n",
    "        print(chemo_name)\n",
    "        \n",
    "        \n",
    "data = data * cell_density * water_content\n",
    "        \n",
    "\n",
    "# convert into counts\n",
    "data = data * 6.022e+23 / 1000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_for_UP2genes = pd.read_csv(f\"{RAW_EXTERNAL}/raw_proteomics_all.csv\")\n",
    "data_for_UP2genes = data_for_UP2genes.drop([0])\n",
    "UP2gene = dict(zip(data_for_UP2genes[\"UP\"], data_for_UP2genes[\"Frist_Gene\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate index uniprot names to gene names and remove enzymes not included in the data \n",
    "data = data.loc[list(set(data.index).intersection(UP2gene.keys()))]\n",
    "data.index = [UP2gene[i] for i in list(set(data.index).intersection(UP2gene.keys()))]\n",
    "#data[\"pyr_1_original\"] = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pyr_1</th>\n",
       "      <th>pyr_2</th>\n",
       "      <th>pyr_3</th>\n",
       "      <th>pyr_original</th>\n",
       "      <th>pyr_original_1</th>\n",
       "      <th>pyr_original_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>fruB</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>135.640649876311</td>\n",
       "      <td>134.115358651721</td>\n",
       "      <td>190.698118029366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yidA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>169.874860452754</td>\n",
       "      <td>163.030636516562</td>\n",
       "      <td>154.619350286062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wzzB</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1107.70800185475</td>\n",
       "      <td>1123.9709840312</td>\n",
       "      <td>1089.71224648648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pdxK</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.739951251735</td>\n",
       "      <td>151.701598573648</td>\n",
       "      <td>119.40983162795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acnA</th>\n",
       "      <td>339091.828617</td>\n",
       "      <td>324000.315521</td>\n",
       "      <td>179672.666517</td>\n",
       "      <td>789.610101001916</td>\n",
       "      <td>765.036028540011</td>\n",
       "      <td>785.354857015051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gutQ</th>\n",
       "      <td>5.419646</td>\n",
       "      <td>5.415451</td>\n",
       "      <td>3.900609</td>\n",
       "      <td>67.9806661287923</td>\n",
       "      <td>60.7404801316662</td>\n",
       "      <td>59.1434971256975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>agp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>908.761562746676</td>\n",
       "      <td>836.462307502063</td>\n",
       "      <td>1237.89732953236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>curA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>480.238188848693</td>\n",
       "      <td>501.380206619107</td>\n",
       "      <td>404.650152086062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yeiR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.7105430910652</td>\n",
       "      <td>24.6931893158493</td>\n",
       "      <td>23.9921762304017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lsrF</th>\n",
       "      <td>3.028778</td>\n",
       "      <td>3.026433</td>\n",
       "      <td>2.179862</td>\n",
       "      <td>105.505789816458</td>\n",
       "      <td>86.8176451946595</td>\n",
       "      <td>108.678926593345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>855 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              pyr_1          pyr_2          pyr_3      pyr_original  \\\n",
       "fruB       0.000000       0.000000       0.000000  135.640649876311   \n",
       "yidA       0.000000       0.000000       0.000000  169.874860452754   \n",
       "wzzB       0.000000       0.000000       0.000000  1107.70800185475   \n",
       "pdxK       0.000000       0.000000       0.000000  120.739951251735   \n",
       "acnA  339091.828617  324000.315521  179672.666517  789.610101001916   \n",
       "...             ...            ...            ...               ...   \n",
       "gutQ       5.419646       5.415451       3.900609  67.9806661287923   \n",
       "agp        0.000000       0.000000       0.000000  908.761562746676   \n",
       "curA       0.000000       0.000000       0.000000  480.238188848693   \n",
       "yeiR       0.000000       0.000000       0.000000  20.7105430910652   \n",
       "lsrF       3.028778       3.026433       2.179862  105.505789816458   \n",
       "\n",
       "        pyr_original_1    pyr_original_2  \n",
       "fruB  134.115358651721  190.698118029366  \n",
       "yidA  163.030636516562  154.619350286062  \n",
       "wzzB   1123.9709840312  1089.71224648648  \n",
       "pdxK  151.701598573648   119.40983162795  \n",
       "acnA  765.036028540011  785.354857015051  \n",
       "...                ...               ...  \n",
       "gutQ  60.7404801316662  59.1434971256975  \n",
       "agp   836.462307502063  1237.89732953236  \n",
       "curA  501.380206619107  404.650152086062  \n",
       "yeiR  24.6931893158493  23.9921762304017  \n",
       "lsrF  86.8176451946595  108.678926593345  \n",
       "\n",
       "[855 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_for_UP2genes.index = data_for_UP2genes['Frist_Gene']\n",
    "\n",
    "data_for_UP2genes = data_for_UP2genes.loc[~data_for_UP2genes.index.duplicated(keep='first')]\n",
    "data = data.loc[~data.index.duplicated(keep='first')]\n",
    "data['ac_original'] = data_for_UP2genes.loc[data.index, ['acetate']]\n",
    "data['ac_original_1'] = data_for_UP2genes.loc[data.index, ['acetate.1']]\n",
    "data['ac_original_2'] = data_for_UP2genes.loc[data.index, ['acetate.2']]\n",
    "\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv(f\"{INTERMEDIATE}/proteomics/pyruvate_usages_counts.csv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
